File: frontend/src/lib/thumbnails.ts
const BASE = "/api/thumbnails";

export interface ThumbnailRenderJob {
  id: string;
  asset_id: string;
  version: number;
  status: "queued" | "processing" | "completed" | "failed";
  provider: string;
  render_type: string;
  output_image: string;
  message: string;
  fallback: boolean;
  created_at: string;
  updated_at: string;
}

export const RENDER_TYPES = [
  { value: "preview", label: "Preview", description: "Standard angled preview shot" },
  { value: "turntable", label: "Turntable", description: "Front-facing rotation preview" },
  { value: "material_preview", label: "Material Preview", description: "Close-up material shot" },
] as const;

export async function renderThumbnail(assetId: string, renderType = "preview"): Promise<ThumbnailRenderJob> {
  const res = await fetch(`${BASE}/render/${assetId}?render_type=${renderType}`, { method: "POST" });
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw new Error(body.detail ?? `Render thumbnail failed (${res.status})`);
  }
  return res.json();
}

export async function captureViewerThumbnail(assetId: string, dataUrl: string): Promise<{ thumbnail_url: string }> {
  const res = await fetch(`${BASE}/capture/${assetId}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ data_url: dataUrl }),
  });
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw new Error(body.detail ?? `Capture thumbnail failed (${res.status})`);
  }
  return res.json();
}

export async function getThumbnailJob(jobId: string): Promise<ThumbnailRenderJob> {
  const res = await fetch(`${BASE}/${jobId}`);
  if (!res.ok) throw new Error(`Get thumbnail job failed (${res.status})`);
  return res.json();
}

export async function listThumbnailJobs(assetId?: string): Promise<ThumbnailRenderJob[]> {
  const url = assetId ? `${BASE}?asset_id=${assetId}` : BASE;
  const res = await fetch(url);
  if (!res.ok) throw new Error(`List thumbnail jobs failed (${res.status})`);
  return res.json();
}

File: backend/app/routes/thumbnails.py
from __future__ import annotations
from typing import Optional
from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

from app.models.thumbnails import ThumbnailRenderJob
from app.services import thumbnail_service

router = APIRouter(prefix="/api/thumbnails", tags=["thumbnails"])


class CaptureRequest(BaseModel):
    data_url: str


@router.post("/render/{asset_id}", response_model=ThumbnailRenderJob, status_code=201)
async def render_thumbnail(asset_id: str, render_type: str = "preview"):
    try:
        job = thumbnail_service.create_thumbnail_job(asset_id, render_type=render_type)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    return job


@router.post("/capture/{asset_id}", status_code=200)
async def capture_viewer_thumbnail(asset_id: str, body: CaptureRequest):
    try:
        url = thumbnail_service.save_viewer_capture(asset_id, body.data_url)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    except Exception as exc:
        raise HTTPException(status_code=500, detail=str(exc))
    return {"thumbnail_url": url}


@router.get("/{job_id}", response_model=ThumbnailRenderJob)
async def get_thumbnail_job(job_id: str):
    job = thumbnail_service.get_thumbnail_job(job_id)
    if job is None:
        raise HTTPException(status_code=404, detail="Thumbnail job not found")
    return job


@router.get("", response_model=list[ThumbnailRenderJob])
async def list_thumbnail_jobs(asset_id: Optional[str] = Query(None)):
    return thumbnail_service.list_thumbnail_jobs(asset_id)
